### PR TITLE
bpo-35121: prefix dot in domain for proper subdomain validation

### PR DESCRIPTION
**Pull request by tirkarthi**
Domain related check is done at https://github.com/python/cpython/blob/0353b4eaaf451ad463ce7eb3074f6b62d332f401/Lib/http/cookiejar.py#L1176 . There is no check to add '.' before domain if absent. Hence it performs a substring match with the values req_host = ".barfoo.com" and erhn = ".barfoo.com" and domain = "foo.com" so the condition `not (req_host.endswith(domain) or erhn.endswith(domain))` fails and doesn't return False. I would suggest adding a check to make sure domain also starts with '.' similar to req_host and erhn thus fixing the issue. I tried the fix and existing tests along with the reported case works fine.

This causes domain "foo.com" to be a valid domain for access of "barfoo.com" since "foo.com" matches the subdomain in substring match with endswith method.

https://bugs.python.org/issue35121

### CODE REVIEW

File: Lib/http/cookiejar.py
Comment (serhiy-storchaka) [inline Lib/http/cookiejar.py]: This will affect calls of `self.is_blocked(domain)` and `self.is_not_allowed(domain)` below.

Comment (tirkarthi) [reply Lib/http/cookiejar.py]: Thanks @serhiy-storchaka . My bad that I looked into fixing the issue and not about the underlying callers that use the dot-prefixed domain. Yes, adding the extra dot makes the comparison to fail where A has an extra dot at start due to my patch at https://github.com/python/cpython/blob/f30060dcd07cd53879226816512ea80bff0d0a78/Lib/http/cookiejar.py#L601 .

Sample program where the domain should be blocked

```python
import urllib
from http.cookiejar import DefaultCookiePolicy

policy = DefaultCookiePolicy(blocked_domains=['xxxfoo.co.jp'])
req = urllib.request.Request('https://xxxfoo.co.jp/')
print(policy.domain_return_ok('xxxfoo.co.jp', req))
```

```
➜  cpython git:(master) ✗ python3.7 /tmp/bar.py 
False
```

```
➜  cpython git:(bpo35121) ✗ ./python.exe /tmp/bar.py  
True
```

With patch this returns true but should be false since the domain is blocked and the prefix dot makes the comparison `.xxxfoo.co.jp == xxxfoo.co.jp` . One fix would be to use dot prefixed domain only for the checks at 

https://github.com/python/cpython/blob/f30060dcd07cd53879226816512ea80bff0d0a78/Lib/http/cookiejar.py#L1178

I think this needs to be fixed but I am also afraid I might accidentally break something here since the function itself received no changes since 2004.

**Comment by tirkarthi** [comment]
Looking further into this the domain validation makes it little more stricter and can have wider implications. Example requests library uses cookiejar to maintain cookies between sessions. One more case is that `domain` can be empty so only non-empty domains can be prefixed with dot

A simple server that sets Cookie with value `A=LDJDSFLKSDJLDSF`

```python
import SimpleHTTPServer
import logging

class MyHTTPRequestHandler(SimpleHTTPServer.SimpleHTTPRequestHandler):
    def do_GET(self):
        self.cookieHeader = self.headers.get('Cookie')
        SimpleHTTPServer.SimpleHTTPRequestHandler.do_GET(self)

    def end_headers(self):
        self.send_my_headers()
        SimpleHTTPServer.SimpleHTTPRequestHandler.end_headers(self)

    def send_my_headers(self):
        self.send_header('Set-Cookie', 'A=LDJDSFLKSDJLDSF')

if __name__ == '__main__':
    SimpleHTTPServer.test(HandlerClass=MyHTTPRequestHandler)

```

Add below host entry to `/etc/hosts` 

```
127.0.0.1 test.com
127.0.0.1 footest.com
```

```python
import requests

with requests.Session() as s:
    cookies = dict(cookies_are='working')
    m = s.get("http://test.com:8000", cookies=cookies)
    print(m.request.headers)
    m = s.get("http://footest.com:8000", cookies=cookies)
    print(m.request.headers)
```

Before patch : 

```
{'User-Agent': 'python-requests/2.11.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 
'Connection': 'keep-alive', 'Cookie': 'cookies_are=working'}
{'User-Agent': 'python-requests/2.11.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 
'Connection': 'keep-alive', 'Cookie': 'A=LDJDSFLKSDJLDSF; cookies_are=working'}
```

After patch :

```
{'User-Agent': 'python-requests/2.11.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 
'Connection': 'keep-alive', 'Cookie': 'cookies_are=working'}
{'User-Agent': 'python-requests/2.11.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 
'Connection': 'keep-alive', 'Cookie': 'cookies_are=working'}
```

As with my patch since the cookie is set on `test.com` while making a request to `footest.com` the cookie is skipped as part of the patch. This is a behavior change to be decided whether worth doing or to document this. A discussion on requests repo on providing the option for user to set a stricter cookie policy : https://github.com/requests/requests/issues/2576

I am adding this to the tracker too.

File: Lib/test/test_http_cookiejar.py
Comment (serhiy-storchaka) [inline Lib/test/test_http_cookiejar.py:430]: Please add also test `("http://foo.bar.com/", "bar.com", True)` above (near line 412 or 417).

File: Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst]: This describes what the code does, which is an implementation detail. A news entry should describe the change in the user visible behavior.

File: Lib/http/cookiejar.py
Comment (ZackerySpytz) [inline Lib/http/cookiejar.py]: New code should conform to PEP 8.

File: Lib/test/test_http_cookiejar.py
Comment (tirkarthi) [reply Lib/test/test_http_cookiejar.py:430]: Added the test at line 418. Thanks.

File: Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst
Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst]: Added a note about it but this is also affects when `domain_return_ok` is present. I couldn't come up with a better wording since I am not a native speaker. Suggestions welcome. This started as a bug fix but since this
turned out to be a security issue is it okay to move this to security section?

```rst
Don't send cookies of domain A without Domain attribute to domain B
when domain A is a suffix match of domain B while using a cookiejar
with :meth:`http.cookiejar.DefaultCookiePolicy` policy. Patch by
Karthikeyan Singaravelan.
```

File: Lib/http/cookiejar.py
Comment (serhiy-storchaka) [reply Lib/http/cookiejar.py]: Maybe use shorter name, e.g. `dotdomain`?

File: Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst]: `:class:`

**Comment by serhiy-storchaka** [comment]
LGTM, but the code style and the wording of a news entry may need some polishing.

File: Lib/http/cookiejar.py
Comment (tirkarthi) [reply Lib/http/cookiejar.py]: `dotdomain` sounds good to me. Perhaps restructure the clause as below removing the assignment at the start?

```python
if domain and not domain.startswith("."):
    dotdomain = "." + domain
else:
    dotdomain = domain
```

**Comment by tirkarthi** [comment]
This also affects 2.7 as I can see the added tests failing without the fix with the same logic being present. I don't know if it might break anything since 2.7 is there for a long time. I have done a manual backport of this PR in case this is needed for 2.7 and locally tests pass for cookielib. Branch : https://github.com/python/cpython/compare/2.7...tirkarthi:bpo35121-27

File: Lib/http/cookiejar.py
Comment (serhiy-storchaka) [reply Lib/http/cookiejar.py]: LGTM.

File: Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst
Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-10-31-15-39-17.bpo-35121.EgHv9k.rst]: Fixed. Thanks.

File: Lib/http/cookiejar.py
Comment (tirkarthi) [reply Lib/http/cookiejar.py]: Changed the code to use `dotdomain` as mentioned in https://github.com/python/cpython/pull/10258#discussion_r243808847. Thanks.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by tirkarthi** [comment]
@serhiy-storchaka Is this worth moving news entry into security section instead of library?

**Comment by tirkarthi** [comment]
Sorry this turned out to be longer since I think I may have found another issue in `return_ok(cookie, domain)` which does similar substring check in `return_ok_domain` thus returning true. So this not only passed cookies of `test.com` to `footest.com` but could have potentially passed cookies of `1.test.com` to `11.test.com`.

I have a fix for `return_ok_domain` locally similar to `domain_return_ok` but I don't know if I am making something stricter though these functions don't return correct values currently as they should be. I would also like to know if it's okay to continue the discussion further here and feedback on the analysis. More details as below : 

https://docs.python.org/3/library/http.cookiejar.html#http.cookiejar.CookiePolicy.return_ok

```
CookiePolicy.return_ok(cookie, request)

    Return boolean value indicating whether cookie should be returned to server.
    cookie is a Cookie instance. request is an object implementing the interface defined by the documentation for CookieJar.add_cookie_header().

```

```python
pol.set_blocked_domains([])
req = urllib.request.Request("http://acme.com/")
res = FakeResponse(headers, "http://acme.com/")
cookies = c.make_cookies(res, req)
c.extract_cookies(res, req)
self.assertEqual(len(c), 1)

print(cookies[0]) # <Cookie CUSTOMER=WILE_E_COYOTE for acme.com/>
req = urllib.request.Request("http://badacme.com/")
self.assertFalse(pol.return_ok(cookies[0], req)) # This fails returning true though cookie is set on acme.com
```

The function return_ok calls helper functions for different attributes of a cookie like return_ok_path, return_ok_domain, etc. In `return_ok_domain` similar substring test is done and thus returns true while checking with domain though it should be checking with dotdomain like the proposed fix at https://github.com/python/cpython/blob/44a79cc5b3d1fb0c03c99077aa26def85ec26c67/Lib/http/cookiejar.py#L1162

From the docs at https://docs.python.org/3/library/http.cookiejar.html#http.cookiejar.CookiePolicy.domain_return_ok

```
 CookiePolicy.domain_return_ok(domain, request)

Return false if cookies should not be returned, given cookie domain.

This method is an optimization. It removes the need for checking every cookie 
with a particular domain (which might involve reading many files). Returning true 
from domain_return_ok() and path_return_ok() leaves all the work to return_ok().
```

`domain_return_ok` is a check that says if the domain has access to the request object as per the original report this should have returned False which it didn't. Then after the lightweight check each cookie itself is checked for all the attributes with `return_ok` as a stricter form of check per cookie which should have returned False with `return_ok_domain` but returned True and thus the cookie for `test.com` was passed along to `footest.com` after two layers of checking. 

Now that `domain_return_ok` is fixed the lightweight check being more stricter at 
https://github.com/python/cpython/blob/44a79cc5b3d1fb0c03c99077aa26def85ec26c67/Lib/http/cookiejar.py#L1251 the domain level returns an empty list but the underlying stricter check `return_ok_domain` still has the bug though it's not executed. The fix would be to use dotdomain in `return_ok_domain` similar to current one. I applied the fix and no tests failed.

One more option is that there are stricter versions of the policy that could be enabled with the [flags](https://docs.python.org/3/library/http.cookiejar.html#http.cookiejar.DefaultCookiePolicy.strict_domain) and hence `return_ok_domain` would have been stricter but the [DefaultCookiePolicy](https://docs.python.org/3/library/http.cookiejar.html#http.cookiejar.DefaultCookiePolicy) has these flags false by default.

**Comment by tirkarthi** [comment]
Fixed `return_ok_domain` too with  b8e2df1 since it's the underlying problem and is used in the public function `return_ok` when domain is set. Let me know if it needs to be reverted and raised a separate PR . Thanks.

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no long accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

**Comment by ned-deily** [comment]
@vstinner, This *is* marked as a security fix.

**Comment by vstinner** [comment]
> @vstinner, This is marked as a security fix.

Sorry. I blindly and manually remove the "needs 3.6 backport" from 57 PRs. I likely made mistakes, but I'm sure that the few changes that must be backported to 3.6 will be updated to add again the label :-)

**Review by ned-deily — COMMENTED** [review]
I'm not sure why I was added as a reviewer as I don't really have any special knowledge or experience in the cookie area so, rather than hold this up more, I will defer to others.

**Comment by tirkarthi** [comment]
Since this is marked as release blocker security issues in bpo with 3.7.3 RC cutoff in few days I will be happy to help if there are any cases that needs to be resolved. I think it would also help if @alex can have an extra review of this if available since they reviewed samesite attribute for `http.cookies` .

Thanks

**Review by serhiy-storchaka — APPROVED** [review]
LGTM again.

Maybe move the news entry to the "Security" section?

**Comment by tirkarthi** [comment]
Thanks @serhiy-storchaka . Moved NEWS entry to security.

**Review by alex — APPROVED** [review]

**Comment by tirkarthi** [comment]
The buildbot failures seem to be unrelated to the issue . I can see open issues for test_ssl https://bugs.python.org/issue35925 and https://bugs.python.org/issue35136 . 

Thanks much Serhiy, Ned, Alex and Zackery for the review and merge.

**Pull request merged**
